File: 04-vector-search/python/Blobs/webDownload.py
# Import necessary modules
from azure.storage.blob import BlobServiceClient
import os.path

# Function to download files from Azure Blob Storage if they don't exist locally
def downloadFilesFromBlobIfTheyDontExist(account_url, container_name, data_folder):
    # Create a BlobServiceClient object which will be used to create a container client
    blob_service_client = BlobServiceClient(account_url=account_url)
    # Get a Container Client using the BlobServiceClient
    container_client = blob_service_client.get_container_client(container_name)

    # List all blobs in the container
    blob_list = container_client.list_blobs()
    for blob in blob_list:
        
        # Construct the file path where the blob will be saved locally
        file_path = data_folder + blob.name

        # Check if the file already exists locally. If not, download it.
        if not os.path.isfile(file_path):
            # Get a Blob Client for the blob
            blob_file = blob_service_client.get_blob_client(container=container_name, blob=blob.name)
            # Open a local file in write mode
            with open(file=file_path, mode='wb') as local_file:
                # Download the blob to a stream
                file_stream = blob_file.download_blob()
                # Write the stream to the local file
                local_file.write(file_stream.readall())

File: 06-guided-project/node.js/Blobs/webDownload.js
// Import the BlobServiceClient from the Azure Storage Blob package
const { BlobServiceClient } = require('@azure/storage-blob');

// Import the file system module
const fs = require('fs');

// Import the path module
const path = require('path');

// Define an asynchronous function to download files from Azure Blob Storage if they don't exist locally
async function downloadFilesFromBlobIfTheyDontExist(accountUrl, containerName, dataFolder) {
    // Create a new BlobServiceClient
    const blobServiceClient = new BlobServiceClient(accountUrl);

    // Get a ContainerClient for the specified container
    const containerClient = blobServiceClient.getContainerClient(containerName);

    // List all blobs in the container
    let blobs = containerClient.listBlobsFlat();

    // Iterate over each blob
    for await (const blob of blobs) {
        // Construct the local file path
        const filePath = path.join(dataFolder, blob.name);

        // Check if the file already exists locally
        if (!fs.existsSync(filePath)) {
            // If the file doesn't exist locally, download it from Azure Blob Storage

            // Get a BlobClient for the blob
            const blobClient = containerClient.getBlobClient(blob.name);

            // Download the blob
            const downloadBlockBlobResponse = await blobClient.download(0);

            // Create a write stream for the local file
            const fileStream = fs.createWriteStream(filePath);

            // Pipe the downloaded blob to the file stream
            downloadBlockBlobResponse.readableStreamBody.pipe(fileStream);
        }
    }
}

// Export the function
module.exports.downloadFilesFromBlobIfTheyDontExist = downloadFilesFromBlobIfTheyDontExist;